### PR TITLE
WIP: Added support for supervisor as process manager

### LINKED ISSUE #693: [13] Use init systems to manage bots and botnet

### PR DESCRIPTION
**Pull request by ondj**
[Supervisor](http://supervisord.org) is process manager written in Python. The main advantage is that it take care about processes, so if bot process exit with failure (exit code different than 0), supervisor try to run it again. Another advantage is that it not require writing PID files. 

## How to use supervisor as process manager in IntelMQ

This manual was tested on Ubuntu 18.04.

Install supervisor. `supervisor_twiddler` is extension for supervisor, that makes possible to create process dynamically. (Ubuntu `supervisor` package is currently based on Python 2, so `supervisor_twiddler` must be installed with Python 2 `pip`.)
```
apt install supervisor python-pip
pip install supervisor_twiddler
```

Create default config `/etc/supervisor/conf.d/intelmq.conf` and restart `supervisor` service:

```ini
[rpcinterface:twiddler]
supervisor.rpcinterface_factory=supervisor_twiddler.rpcinterface:make_twiddler_rpcinterface

[group:intelmq]
```

Change IntelMQ process manager in `/opt/intelmq/etc/defaults.conf`:

```
"proccess_manager": "supervisor",
```

After this is possible to manage bots like before with `intelmqctl` command.

Fixes #693

### CODE REVIEW

**Comment by wagner-certat** [comment]
Hi,

Thanks for the work! I really appreciate this (and in general all your) contribution(s). As I am currently traveling a lot, I wasn't able to review it yet. I hope I can do it before Christmas, but I can't promise it unfortunately.

File: intelmq/bin/intelmqctl.py [hunk 1]
```diff
@@ -346,7 +349,237 @@ def __status_process(self, pid, module, bot_id):
...
+                log_bot_error("stopped", bot_id)
+                return "stopped"
+
+        if state not in (self.ProcessState.RUNNING, self.ProcessState.STARTING, self.ProcessState.BACKOFF):
+            self._remove_bot(bot_id)
+            log_bot_error("stopped", bot_id)
+            return "stopped"
+
+        log_bot_message("reloading", bot_id)
+        self._get_supervisor().supervisor.signalProcess(self._process_name(bot_id), "HUP")
```
Comment (wagner-certat) [inline intelmq/bin/intelmqctl.py]: Reloading did not work:
```
$ intelmqctl reload abusech-feodo-domains-collector
intelmqctl: Connected to supervisor 'supervisor' (API version 3.0)
intelmqctl: Reloading bot abusech-feodo-domains-collector ...
Traceback (most recent call last):
  File "/usr/local/bin/intelmqctl", line 11, in <module>
    load_entry_point('intelmq', 'console_scripts', 'intelmqctl')()
  File "/home/sebastianw/intelmq/intelmq/bin/intelmqctl.py", line 1377, in main
    return x.run()
  File "/home/sebastianw/intelmq/intelmq/bin/intelmqctl.py", line 869, in run
    retval, results = args.func(**args_dict)
  File "/home/sebastianw/intelmq/intelmq/bin/intelmqctl.py", line 902, in bot_reload
    status = self.bot_process_manager.bot_reload(bot_id, getstatus)
  File "/home/sebastianw/intelmq/intelmq/bin/intelmqctl.py", line 478, in bot_reload
    self._get_supervisor().supervisor.signalProcess(self._process_name(bot_id), "HUP")
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1089, in __call__
    return self.__send(self.__name, args)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1419, in __request
    verbose=self.__verbose
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1131, in request
    return self.single_request(host, handler, request_body, verbose)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1147, in single_request
    return self.parse_response(resp)
  File "/usr/lib/python3.4/xmlrpc/client.py", line 1318, in parse_response
    return u.close()
  File "/usr/lib/python3.4/xmlrpc/client.py", line 656, in close
    raise Fault(**self._stack[0])
xmlrpc.client.Fault: <Fault 1: 'UNKNOWN_METHOD'>
```

That has been added in 3.2.0 but I have 3.0.0 (Debian jessie): http://supervisord.org/changes.html#id11
Maybe we can catch this?

File: intelmq/bin/intelmqctl.py [hunk 2]
```diff
@@ -346,7 +349,237 @@ def __status_process(self, pid, module, bot_id):
...
+            if supervisor_state["statename"] != "RUNNING":
+                raise Exception("Unexpected supervisor state {}".format(supervisor_state["statename"]))
+
+            try:
+                self.__supervisor_xmlrpc.twiddler.getAPIVersion()
+            except xmlrpc.client.Fault:
+                raise Exception("Twiddler is not supported. Is Twiddler for supervisor installed and enabled?")
+
+            if self.SUPERVISOR_GROUP not in self.__supervisor_xmlrpc.twiddler.getGroupNames():
+                raise Exception("Supervisor does not contains process group {}. It must be created manually.")
```
Comment (wagner-certat) [inline intelmq/bin/intelmqctl.py]: Looks like you forgot to format that string

File: intelmq/bin/intelmqctl.py [hunk 3]
```diff
@@ -346,7 +349,237 @@ def __status_process(self, pid, module, bot_id):
...
+
+            supervisor_state = self.__supervisor_xmlrpc.supervisor.getState()
+
+            if supervisor_state["statename"] != "RUNNING":
+                raise Exception("Unexpected supervisor state {}".format(supervisor_state["statename"]))
+
+            try:
+                self.__supervisor_xmlrpc.twiddler.getAPIVersion()
+            except xmlrpc.client.Fault:
+                raise Exception("Twiddler is not supported. Is Twiddler for supervisor installed and enabled?")
```
[the remark this answers is not in the document]
Comment (ondj) [reply intelmq/bin/intelmqctl.py]: Fixed, thanks. Code now uses `IntelMQController.abort` method.

File: intelmq/bin/intelmqctl.py (hunk 1 above)
Comment (ondj) [reply intelmq/bin/intelmqctl.py]: I added better error message and also warning when old supervisor version is used.

File: intelmq/bin/intelmqctl.py (hunk 2 above)
Comment (ondj) [reply intelmq/bin/intelmqctl.py]: Yes, fixed.

**Comment by ondj** [comment]
I fixed your suggestion, the error message for not writable socket is more clear now, but still without suggested solution – because I think there can be others problems too (for example with selinux). 

But this code is still WIP, because of bots runs under root user. It is possible to set that bots will run under intelmq user, but still attacker with intelmq user can run process under root user. So now I am trying to find solution for that problem.

**Comment by wagner-certat** [comment]
All bots are changing to the unprivileged intelmq user now themselves.

**Comment by wagner-certat** [comment]
I now merged it so we can get some more feedback or support from other users and developers. In the docs I clearly marked it as beta

**Pull request merged**